rocky/reports/report_types/name_server_report/report.py

```diff
@@ -125,3 +126,46 @@ def generate_data(self, input_ooi: str, valid_time: datetime) -> dict[str, Any]:
             "name_server_checks": name_server_checks,
             "finding_types": sorted(finding_types.values(), reverse=True, key=lambda x: x.risk_severity),
         }
+
+    def collect_data(self, input_oois: Iterable[str], valid_time: datetime) -> dict[str, dict[str, Any]]:
+        hostnames_by_input_ooi = self.to_hostnames(input_oois, valid_time)
+        all_hostnames = [h for key, hostnames in hostnames_by_input_ooi.items() for h in hostnames]
+
+        query = "Hostname.<ooi[is Finding].finding_type"
+        hostname_finding_types = self.group_by_source(
+            self.octopoes_api_connector.query_many(query, valid_time, all_hostnames),
+            lambda ooi: ooi.id in ["KAT-NO-DNSSEC", "KAT-INVALID-DNSSEC"],
+        )
+        query = "Hostname.<hostname[is ResolvedHostname].address.<address[is IPPort].<ooi[is Finding].finding_type"
+        port_finding_types = self.group_by_source(
+            self.octopoes_api_connector.query_many(query, valid_time, all_hostnames),
+            lambda ooi: ooi.id in ["KAT-UNCOMMON-OPEN-PORT", "KAT-OPEN-SYSADMIN-PORT", "KAT-OPEN-DATABASE-PORT"],
+        )
+
+        result = {}
+
+        for input_ooi, hostname_references in hostnames_by_input_ooi.items():
+            name_server_checks = NameServerChecks()
+            finding_types = {}
+
+            for hostname in hostname_references:
+                check = NameServerCheck()
+                check.no_uncommon_ports = not any(port_finding_types.get(hostname, []))
+                check.has_dnssec = "KAT-NO-DNSSEC" not in [x.id for x in hostname_finding_types.get(hostname, [])]
```

Comment:
Use a generator instead, this will improve performance in this case
```suggestion
                check.has_dnssec = "KAT-NO-DNSSEC" not in (x.id for x in hostname_finding_types.get(hostname, []))
```

Reply:
Is this really worth the effort? If I only change it here this probably will only introduce inconsistencies as most of our comprehensions are just list comprehensions and the underlying data structure is also just a plain list (so we do not save memory)

Reply:
Yep it's absolutely worth the effort. Because this optimization not only saves memory, it also enhances performance by minimizing the number of iterations. There's no need to create a new list based on another list and iterate over each list element if there's a greater chance that the condition (not in) is met earlier and just to throw the list away after it. But even if it's at the end of the list (because it's still an O(n) search), it will still be beneficial over creating a new list and disposing it.

Of course this should be applied on more places for consistency and overall readability and performance optimization, but this was just an idea and obviously I can't pin point every other example where this is applicable.